src/wasm2asm.h

```diff
@@ -540,26 +568,53 @@ Ref Wasm2AsmBuilder::processFunction(Function* func) {
   ret[3]->push_back(theVar);
   // body
   scanFunctionBody(func->body);
-  if (isStatement(func->body)) {
-    IString result =
-        func->result != none ? getTemp(func->result, func) : NO_RESULT;
-    flattenAppend(ret, ValueBuilder::makeStatement(processFunctionBody(func, result)));
-    if (func->result != none) {
-      // do the actual return
-      ret[3]->push_back(ValueBuilder::makeStatement(ValueBuilder::makeReturn(makeAsmCoercion(ValueBuilder::makeName(result), wasmToAsmType(func->result)))));
-      freeTemp(func->result, result);
-    }
+  bool isBodyBlock = (func->body->_id == Expression::BlockId);
+  ExpressionList* stats = isBodyBlock ?
+      &static_cast<Block*>(func->body)->list : nullptr;
+  bool endsInReturn =
+      (isBodyBlock && ((*stats)[stats->size()-1]->_id == Expression::ReturnId));
+  if (endsInReturn) {
+    flattenAppend(ret, processFunctionBody(func, NO_RESULT));
   } else {
-    // whole thing is an expression, just do a return
-    if (func->result != none) {
-      ret[3]->push_back(ValueBuilder::makeStatement(ValueBuilder::makeReturn(makeAsmCoercion(processFunctionBody(func, EXPRESSION_RESULT), wasmToAsmType(func->result)))));
+    if (isStatement(func->body)) {
+      IString result =
+          func->result != none ? getTemp(func->result, func) : NO_RESULT;
+      flattenAppend(ret, processFunctionBody(func, result));
+      if (func->result != none) {
+        // do the actual return
+        ret[3]->push_back(
+          ValueBuilder::makeReturn(
+            makeAsmCoercion(
+              ValueBuilder::makeName(result),
+              wasmToAsmType(func->result)
+            )
+          )
+        );
+        freeTemp(func->result, result);
+      }
     } else {
-      flattenAppend(ret, processFunctionBody(func, NO_RESULT));
+      // whole thing is an expression, just do a return
+      if (func->result != none) {
+        ret[3]->push_back(
```

Comment:
This gets reused almost wholesale from just above. Helper function? Possibly helper lambda?